### PR TITLE
Ft user registration login endpoint

### PR DESCRIPTION
**Pull request by NgangaMaryanne**
#### What does this PR do?
Updates models and endpoint to allow token based authentication.

#### Description of Task to be completed?
A user should be able to login and get a token that can be used to access other resources instead of sending login credentials all the time. 
#### How can this be manually tested
* Clone repository
* Test endpoint `localhost/auth/register` and `localhost/auth/login` on postman.

#### Any background context you want to provide?
Pyjwt which is the package that has been used to generate the tokens that are sent to the user during login and the user uses the tokens for subsequent access to the API resources.

### CODE REVIEW

File: bucketlist/__init__.py [hunk 1]
```diff
@@ -1,15 +1,31 @@
...
+from config import app_config
 db = SQLAlchemy()
 login_manager = LoginManager()
 
+
 def create_app(config_name):
-    app = Flask(__name__)
+    app = Flask(__name__, instance_relative_config=True)
+    app.config.from_object(app_config[config_name])
+    # app.config.from_pyfile('config.py')
```
Comment (gacheruevans) [inline bucketlist/__init__.py]: remove commented code

File: bucketlist/auth/views.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        parser = reqparse.RequestParser()
+        parser.add_argument('email', required=True, help='email is required')
+        parser.add_argument('username', required=True, help='Username is required')
+        parser.add_argument('first_name')
+        parser.add_argument('last_name')
+        parser.add_argument('password')
+        user_data = parser.parse_args(strict=True)
+        if user_data:
+            #check that email is entered correctly.
+            if not re.match(r"[^@]+@[^@]+\.[^@]+", user_data['email']):
```
Comment (gacheruevans) [inline bucketlist/auth/views.py]: good email validation, but can it validate **maryanne@andela..com**

File: bucketlist/auth/views.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                return make_response(jsonify(response))
+            #password has to be more than 8 characters.    
+            if len(user_data['password'])<8:
+                response = {'message':'Password too short. Please try again'}
+                return make_response(jsonify(response))
+                #check that username is unique
+            user_usrname = user = User.query.filter_by(username=user_data['username']).first()
+            if user_usrname:
+                response = {'message':'Username already exists. Please try again'}
+                return make_response(jsonify(response))
```
Comment (gacheruevans) [inline bucketlist/auth/views.py]: add status codes

File: bucketlist/auth/views.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                user=User.query.filter_by(email=user_details['email']).first()
+                # auth_token = user.encode_auth_token(user.id)
+                if user and user.verify_password(user_details['password']):
+                    auth_token = user.encode_auth_token(user.id)
+                    if auth_token:
+                        #set global
+                        response = {'status':'success',
+                                    'message':'Successfully logged in.',
+                                    'auth_token':auth_token.decode()}
+                        return make_response(jsonify(response))
```
Comment (gacheruevans) [inline bucketlist/auth/views.py]: add page status code

File: bucketlist/auth/views.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                            'error':str(Exception)}
+                return make_response(jsonify(response))
+        else:
+            response = {'status':'fail',
+                        'message':'Please input details.'}
+        #add edge cases for wrong email or wrong password.
+
+class UserStatus(Resource):
+    def get(self):
+        auth_header = request.headers.get('Authorization')
```
Comment (gacheruevans) [inline bucketlist/auth/views.py]: Add comments that state the purpose of the function

File: bucketlist/models.py [hunk 6]
```diff
@@ -30,36 +44,101 @@ def password(self, password):
...
+            payload = {
+            'exp':datetime.datetime.utcnow()+datetime.timedelta(days=1),
+            'iat':datetime.datetime.utcnow(),
+            'sub':user_id
+            }
+            return jwt.encode(payload, os.getenv('SECRET_KEY'), algorithm='HS256')
+        except(Exception):
+            return Exception
+
+
```
Comment (gacheruevans) [inline bucketlist/models.py]: handle your white spaces

File: bucketlist/__init__.py (hunk 1 above)
Comment (NgangaMaryanne) [reply bucketlist/__init__.py]: done

File: bucketlist/auth/views.py (hunk 2 above)
Comment (NgangaMaryanne) [reply bucketlist/auth/views.py]: will check that.

**Pull request merged**